src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs

```diff
@@ -174,5 +214,143 @@ private static bool TryParseArguments(string[] args, out (Dictionary<string, str
+        {
+            errorWriter.WriteLine(format, args);
+
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to deserialize global properties from the standard input stream.
+        /// </summary>
+        /// <remarks>
+        /// The stream backed by the specified <see cref="BinaryReader" /> may contain a preamble/byte order mark.   Preambles are variable length
+        /// from 2 to 4 bytes.  The first 4 bytes are either:
+        ///   -Variable length preamble and any remaining bytes of the actual content
+        ///   -No preamble and just the integer representing the number of items in the dictionary
+        ///
+        /// Since the preamble could be 3 bytes, that means that the last byte in the buffer will be the first byte of the next segment. So this code
+        /// determines how long the preamble is, replaces the remaining bytes at the beginning of the buffer, and copies the next set of bytes.  This
+        /// effectively "skips" the preamble by eventually having the buffer contain the first 4 bytes of the content that should actually be read.
+        ///
+        /// Example stream:
+        /// 
+        /// |   3 byte preamble  |      4 byte integer       |
+        /// |------|------|------|------|------|------|------|
+        /// | 0xFF | 0XEF | 0x00 | 0x07 | 0x00 | 0x00 | 0x00 |
+        ///
+        /// The first 4 bytes are read into the buffer (notice one of the bytes is actually not the preamble):
+        /// 
+        /// |       4 byte buffer       |
+        /// |------|------|------|------|
+        /// | 0xFF | 0XEF | 0x00 | 0x07 |
+        ///
+        /// If the first three bytes match the preamble (0xFF, 0xEF, 0x00), then the array is shifted so that last item becomes the first:
+        /// 
+        /// |       4 byte buffer       |
+        /// |------|------|------|------|
+        /// | 0x07 | 0XEF | 0x00 | 0x07 |
+        ///
+        /// Since only 1 byte was moved up in the buffer, then the next 3 bytes from the stream are read:
+        ///
+        /// |       4 byte buffer       |
+        /// |------|------|------|------|
+        /// | 0x07 | 0X00 | 0x00 | 0x00 |
+        ///
+        /// Now the buffer contains the integer in the stream and the preamble is skipped.
+        /// 
+        /// </remarks>
+        /// <param name="errorWriter">A <see cref="TextWriter" /> to write errors to if one occurs.</param>
+        /// <param name="reader">The <see cref="BinaryReader" /> to use when deserializing the arguments.</param>
+        /// <param name="globalProperties">Receives a <see cref="Dictionary{TKey, TValue}" /> representing the global properties.</param>
+        /// <returns><see langword="true" /> if the arguments were successfully deserialized, otherwise <see langword="false" />.</returns>
+        internal static bool TryDeserializeGlobalProperties(TextWriter errorWriter, BinaryReader reader, out Dictionary<string, string> globalProperties)
```

Comment:
I know you already spent a lot of time investigating, and I'm sorry if this is something you already tried.

I tried putting a `LogError(errorWriter, "Preable length: " + preable.Length)` just after the `GetPreamble()` code, and confirmed on my machine that it expects a preamble.  By changing `detectEncodingFromByteOrderMarks` from `true` to `false`, the preamble length becomes zero, but obviously this doesn't solve the problem that the preamble was being written to the StdIn stream, therefore breaking the binary parser.

What I couldn't figure out to verify myself, is in `StaticGraphRestoreTaskBase.cs`, where the `ProcessStartInfo` is created, create your own instance of a UTF8 encoder, explicitly disabling the "identifier" (I'm guessing this is the preamble): `var encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);`, then set `startInfo.StandardInputEncoding = encoding`. Probably should set it on `startInfo.StandardOutputEncoding`, so the encoding both ways is deterministic. It looks like you pass an encoding to `BinaryWriter`'s constructor too, so you probably need to set it there as well. In static restore's `Program.Main`, similarly set `System.Console.InputEncoding = System.Console.OutputEncoding = new UTF8Encoding(false);`, although this might not be necessary for input encoding, since it's using a binary reader.

It's like when writing text files with .NET, the default encoding writes what text editors call the Byte Order Marker (BOM), which I guess is this preamble. Set the StreamWriter's encoding instance to `new UTF8Encoding(false)`, and it should no longer write that BOM at the beginning of the file. By creating your own instance of UTF8Encoding and telling it not to write it, I hope we don't need the complex preable code in static graph restore's code at all.

Having written all that, I also think that using binary writer is unnecessary, and I think is causing a lot of the fragility. If the global properties are written to StdIn as a string, the CPU cost to parse it is not huge. When the global properties are passed on the command line, you're processing them as a string, so I don't think the StdIn parser needs to go from one extreme (string & LINQ heavy, so many string allocations) to the other (binary). We can use text over StdIn as a middle ground, resolving the command line too long problem (and maybe do it in a way that doesn't need so many allocations and string splitting), while avoiding the preamble breaking binary parsing issue.

Reply:
After spending most of the day trying to figure out how to test this child process stuff, I think I finally verified that `new UTF8Encoding(false)` does not work 😞 

Oh well.  I still think a `while ((line = Console.ReadLine()) != null)` loop would be less fragile and simpler code. Better "startup" performance than the string.Split + LINQ code used for `args[3]` as well.

Reply:
I don't think we can rely on setting the encoding of the stream, the docs all but say its unpredictable.  So I think we'll always have to harden against it.  I lean towards binary serialization assuming its very fast, here are some benchmarks I did where my code is serializing/deserializing over STDIN.  I don't think the memory allocations count here since the deserialization is happening in another process.

| Method                  | PropertyCount | Mean      | Error    | StdDev   | Allocated |
|------------------------ |-------------- |----------:|---------:|---------:|----------:|
| BenchmarkBinary         | 10            |  80.35 ms | 0.508 ms | 0.450 ms |  14.13 KB |
| BenchmarkNewtonsoftJson | 10            | 138.74 ms | 1.365 ms | 1.210 ms |  14.83 KB |
| BenchmarkSystemTextJson | 10            | 109.58 ms | 2.127 ms | 2.276 ms |  14.33 KB |
| BenchmarkBinary         | 100           |  82.20 ms | 1.148 ms | 1.018 ms |  14.14 KB |
| BenchmarkNewtonsoftJson | 100           | 146.30 ms | 2.922 ms | 4.882 ms |  14.83 KB |
| BenchmarkSystemTextJson | 100           | 106.27 ms | 1.559 ms | 1.458 ms |  14.33 KB |
| BenchmarkBinary         | 1000          |  86.59 ms | 1.053 ms | 0.985 ms |  14.14 KB |
| BenchmarkNewtonsoftJson | 1000          | 158.83 ms | 2.744 ms | 3.266 ms |  14.83 KB |
| BenchmarkSystemTextJson | 1000          | 107.20 ms | 1.637 ms | 1.367 ms |  14.49 KB |

```c#
internal static Dictionary<string, string>? SystemTextJsonDeserialize(Stream stream)
{
    return JsonSerializer.Deserialize<Dictionary<string, string>>(stream);
}

internal static void SystemTextJsonSerialize(Process process, Dictionary<string, string> properties)
{
    JsonSerializer.Serialize(process.StandardInput.BaseStream, properties);

    process.StandardInput.Close();
}

internal static Dictionary<string, string>? NewtonsoftJsonDeserialize(Stream stream)
{
    StreamReader streamReader = new(stream);

    using JsonTextReader jsonReader = new(streamReader);

    NewtonsoftJsonSerializer serializer = new();

    return serializer.Deserialize<Dictionary<string, string>>(jsonReader);
}

internal static void NewtonsoftJsonSerialize(Process process, Dictionary<string, string> properties)
{
    NewtonsoftJsonSerializer serializer = new();

    serializer.Serialize(process.StandardInput, properties);

    process.StandardInput.Flush();
}

internal static void BinarySerialize(Process process, Dictionary<string, string> properties)
{
    using BinaryWriter writer = new(process.StandardInput.BaseStream);

    writer.Write(properties.Count);

    foreach (KeyValuePair<string, string> option in properties)
    {
        writer.Write(option.Key);
        writer.Write(option.Value);
    }
}

internal static Dictionary<string, string>? BinaryDeserialize(Stream stream)
{
    // Long method that detects and skips the preamble
}
```

Both `System.Text.Json` and `Newtonsoft.Json` appear to be able to skip the preamble/BOM which is nice.

Reply:
System.Text.Json is unapologetically UTF8 only:
* https://learn.microsoft.com/en-us/dotnet/api/system.text.json?view=net-7.0
* https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserialize?view=net-7.0#system-text-json-jsonserializer-deserialize(system-io-stream-system-type-system-text-json-jsonserializeroptions)


Newtonsoft.Json works over `TextReader` (via `JsonTextReader`), so it should work as long the StreamReader that is `System.Console.StandardInput` correctly detects the encoding.

This is why I thought explicitly setting the encoding on `ProcessStartInfo.InputEncoding` would work. If it did what it sounds like it's supposed to do, it sounds like it should be possible to prevent the BOM/preamble from being written 😞

Reply:
Well this is interesting, serializing everything as a string is faster?  This code writes an integer, then a key, then the value, then a unicode `Ux0004` to mark the end of a property since MSBuild property values can have line breaks.

```c#
internal static void StringSerialize(Process process, Dictionary<string, string> properties)
{
    process.StandardInput.WriteLine(properties.Count);

    foreach(KeyValuePair<string, string> property in properties)
    {
        process.StandardInput.WriteLine(property.Key);
        process.StandardInput.WriteLine(property.Value);
        process.StandardInput.WriteLine((char)4);
    }

    process.StandardInput.Close();
}

internal static Dictionary<string, string> StringDeserialize(Stream stream)
{
    using StreamReader reader = new(stream);

    string? countString = reader.ReadLine();

    Dictionary<string, string> properties = new();

    for (int i = 0; i < int.Parse(countString); i++)
    {
        string? key = reader.ReadLine();
        string? value = reader.ReadLine();

        StringBuilder sb = new StringBuilder(value);

        value = reader.ReadLine();

        while(value != ((char)4).ToString())
        {
            sb.AppendLine();
            sb.Append(value);

            value = reader.ReadLine();
        }

        properties[key] = sb.ToString();
    }

    return properties;
}
```

| Method                  | PropertyCount | Mean      | Error    | StdDev   | Allocated |
|------------------------ |-------------- |----------:|---------:|---------:|----------:|
| BenchmarkBinary         | 10            |  83.49 ms | 0.645 ms | 0.603 ms |  15.01 KB |
| BenchmarkNewtonsoftJson | 10            | 142.04 ms | 2.279 ms | 2.132 ms |  16.14 KB |
| BenchmarkSystemTextJson | 10            | 107.37 ms | 0.997 ms | 0.884 ms |  15.37 KB |
| BenchmarkString         | 10            |  82.24 ms | 1.028 ms | 0.962 ms |  14.86 KB |
| BenchmarkBinary         | 100           |  83.02 ms | 0.786 ms | 0.656 ms |  14.13 KB |
| BenchmarkNewtonsoftJson | 100           | 141.50 ms | 0.864 ms | 0.722 ms |  16.14 KB |
| BenchmarkSystemTextJson | 100           | 108.07 ms | 0.660 ms | 0.617 ms |  14.32 KB |
| BenchmarkString         | 100           |  81.42 ms | 0.832 ms | 0.778 ms |  14.11 KB |
| BenchmarkBinary         | 1000          |  87.46 ms | 1.143 ms | 1.013 ms |  15.01 KB |
| BenchmarkNewtonsoftJson | 1000          | 160.87 ms | 1.535 ms | 1.360 ms |  14.82 KB |
| BenchmarkSystemTextJson | 1000          | 108.42 ms | 0.760 ms | 0.635 ms |  15.48 KB |
| BenchmarkString         | 1000          |  89.02 ms | 0.935 ms | 0.780 ms |     15 KB |

Reply:
This is now moot, the problem is fixed by setting `Console.InputEncoding` when creating the process.